### PR TITLE
Use locations JSON to reattach to instances based on roles

### PR DESCRIPTION
**Pull request by whoarethebritons**

### CODE REVIEW

**Comment by whoarethebritons** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/2701/

File: appscale/tools/node_layout.py [hunk 1]
```diff
@@ -709,6 +709,42 @@ def to_list(self):
     return [node.to_json() for node in self.nodes]
 
 
+  def from_locations_json_list(self, locations_nodes_list):
+    """Returns a list of nodes if the previous locations JSON matches with the
+    current NodeLayout from the AppScalefile. Otherwise returns None."""
+
+    # If the length does not match up the user has added a node in the
```
Comment (menivaitsi) [inline appscale/tools/node_layout.py]: Or removed?

File: appscale/tools/node_layout.py [hunk 2]
```diff
@@ -856,6 +892,23 @@ def to_json(self):
     }
 
 
+  def from_json(self, node_dict):
```
Comment (menivaitsi) [inline appscale/tools/node_layout.py]: Add docstring.

File: appscale/tools/remote_helper.py [hunk 3]
```diff
@@ -131,14 +131,30 @@ def start_all_nodes(cls, options, node_layout):
...
-      # Set newly obtained node layout info for this deployment.
-
-      for i, sorted_instance_id in enumerate(sorted(instance_ids)):
-        index = instance_ids.index(sorted_instance_id)
+      # Get the node_info from the locations JSON.
+      node_info = LocalState.get_local_nodes_info(keyname=options.keyname)
+
+      previous_node_list = node_layout.from_locations_json_list(node_info)
+      # If this is None, the AppScalefile has been changed or the nodes could
+      #  not be matched up by roles/jobs.
```
Comment (menivaitsi) [inline appscale/tools/remote_helper.py]: Please remove space extra space before "not".

File: appscale/tools/remote_helper.py [hunk 4]
```diff
@@ -131,14 +131,30 @@ def start_all_nodes(cls, options, node_layout):
...
+      if previous_node_list is None:
+        raise BadConfigurationException("AppScale does not currently support "
+                                        "changes to AppScalefile or locations "
+                                        "JSON between a down and an up. If "
+                                        "you would like to "
+                                        "change the node layout use "
+                                        "down --terminate before an up.")
+      node_layout.nodes = previous_node_list
+      for i, node in enumerate(node_layout.nodes):
```
Comment (menivaitsi) [inline appscale/tools/remote_helper.py]: Please refrain from using single-letter variable names. You can use something like: _node_index_

File: appscale/tools/remote_helper.py [hunk 5]
```diff
@@ -148,6 +164,12 @@ def start_all_nodes(cls, options, node_layout):
       options, agent, params,
       len(load_balancer_nodes))
 
+    for i, node in enumerate(load_balancer_nodes):
```
Comment (menivaitsi) [inline appscale/tools/remote_helper.py]: Same here.

File: appscale/tools/remote_helper.py [hunk 6]
```diff
@@ -156,15 +178,12 @@ def start_all_nodes(cls, options, node_layout):
...
-      public_ips.extend(_public_ips)
-      private_ips.extend(_private_ips)
-
-    for i, sorted_instance_id in enumerate(sorted(instance_ids)):
-      index = instance_ids.index(sorted_instance_id)
-      node_layout.nodes[i].public_ip = public_ips[index]
-      node_layout.nodes[i].private_ip = private_ips[index]
-      node_layout.nodes[i].instance_id = instance_ids[index]
+
+      for i, node in enumerate(other_nodes):
```
Comment (menivaitsi) [inline appscale/tools/remote_helper.py]: And here.

File: test/test_appscale_run_instances.py [hunk 7]
```diff
@@ -731,6 +731,25 @@ def test_appscale_in_one_node_cloud_deployment_manual_spot_price(self):
...
+    # Let's mock the call to describe_instances when checking for old
+    # instances to re-use, and then to start the headnode.
+    pending_instance = flexmock(name='pending_instance', state='pending',
+                                key_name=self.keyname, id='i-ABCDEFG')
+    pending_reservation = flexmock(name='pending_reservation',
+                                   instances=[pending_instance])
+
+    no_instances = flexmock(name='no_instances', instances=[])
+    running_instance = flexmock(name='running_instance', state='running',
+                                key_name=self.keyname, id='i-ABCDEFG', ip_address='public1',
```
Comment (menivaitsi) [inline test/test_appscale_run_instances.py]: This line exceeds 80 chars, please break it up.

File: test/test_appscale_terminate_instances.py [hunk 8]
```diff
@@ -303,6 +303,119 @@ def test_terminate_in_cloud_and_succeeds(self):
     AppScaleTools.terminate_instances(options)
 
 
+  def test_terminate_in_cloud_with_unreachable_node(self):
+    # let's say that there is a locations.json file with key
```
Comment (menivaitsi) [inline test/test_appscale_terminate_instances.py]: Add docstring.

File: test/test_appscale_terminate_instances.py [hunk 9]
```diff
@@ -303,6 +303,119 @@ def test_terminate_in_cloud_and_succeeds(self):
...
+    # 'infrastructure_info', which means appscale is running, so we should
+    # terminate the services on each box
```
Comment (menivaitsi) [inline test/test_appscale_terminate_instances.py]: Start this sentence with upper case and end with period.

File: test/test_appscale_terminate_instances.py [hunk 10]
```diff
@@ -303,6 +303,119 @@ def test_terminate_in_cloud_and_succeeds(self):
...
+    flexmock(os.path)
+    os.path.should_call('exists')  # set up the fall-through
+    os.path.should_receive('exists').with_args(
+      LocalState.get_secret_key_location(self.keyname)).and_return(True)
+
+    # mock out reading the locations.json file with key
```
Comment (menivaitsi) [inline test/test_appscale_terminate_instances.py]: Same throughout.

File: test/test_appscale_terminate_instances.py [hunk 11]
```diff
@@ -303,6 +303,119 @@ def test_terminate_in_cloud_and_succeeds(self):
...
+    # our deployment
+    fake_one_running = flexmock(name='fake_one', key_name=self.keyname, state='running',
+                                id='i-ONE', ip_address='1.2.3.4', private_ip_address='1.2.3.4')
+    fake_two_running = flexmock(name='fake_two', key_name=self.keyname, state='running',
+                                id='i-TWO', ip_address='1.2.3.4', private_ip_address='1.2.3.4')
+    fake_three_running = flexmock(name='fake_three', key_name='abcdefg',
+                                  state='running', id='i-THREE', ip_address='1.2.3.4',
+                                  private_ip_address='1.2.3.4')
+    fake_reservation_running = flexmock(name='fake_reservation', instances=[fake_one_running,
+                                                                            fake_two_running, fake_three_running])
```
Comment (menivaitsi) [inline test/test_appscale_terminate_instances.py]: Line too long. Please make a pass and stick to 80 chars.

File: test/test_appscale_terminate_instances.py (hunk 8 above)
Comment (whoarethebritons) [inline test/test_appscale_terminate_instances.py]: This file was not meant to be committed... I'm going to remove the changes from this PR.

File: appscale/tools/node_layout.py (hunk 2 above)
Comment (whoarethebritons) [inline appscale/tools/node_layout.py]: I'm not sure why this isn't being marked as done.

**Pull request merged**